### PR TITLE
XMark Benchmark Documentation + static push to meta data provider

### PR DESCRIPTION
**Pull request by shivanimall**

### CODE REVIEW

File: vxquery-core/src/main/java/org/apache/vxquery/xmlquery/query/XMLQueryCompiler.java [hunk 1]
```diff
@@ -164,13 +164,7 @@ public Object getType(ILogicalExpression expr, IMetadataProvider<?, ?> metadataP
...
             @Override
             public Object makeNullableType(Object type) throws AlgebricksException {
-                SequenceType st = (SequenceType) type;
-                if (st.getQuantifier().allowsEmptySequence()) {
-                    return type;
-                } else if (st.getQuantifier() == Quantifier.QUANT_ONE) {
-                    return SequenceType.create(st.getItemType(), Quantifier.QUANT_QUESTION);
-                }
-                return type;
+                throw new NotImplementedException("NullableTypeComputer is not implented (makeNullableType)");
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/xmlquery/query/XMLQueryCompiler.java]: Do not remove this from master. This change is needed by the inclusion of new rewrite rules.

File: vxquery-core/src/main/java/org/apache/vxquery/xmlquery/query/XMLQueryCompiler.java [hunk 2]
```diff
@@ -245,8 +240,6 @@ public Module getModule() {
                 RewriteRuleset.buildNestedDataSourceRuleCollection()));
         defaultLogicalRewrites.add(new Pair<AbstractRuleController, List<IAlgebraicRewriteRule>>(seqOnceCtrl,
                 RewriteRuleset.buildTypeInferenceRuleCollection()));
-        defaultLogicalRewrites.add(new Pair<AbstractRuleController, List<IAlgebraicRewriteRule>>(seqCtrlNoDfs,
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/xmlquery/query/XMLQueryCompiler.java:248]: Do not remove this from master. This change is needed by the inclusion of new rewrite rules.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushChildIntoDataScanRule.java [hunk 3]
```diff
@@ -72,13 +66,17 @@
...
 public class PushChildIntoDataScanRule extends AbstractUsedVariablesProcessingRule {
-    final StaticContextImpl dCtx = new StaticContextImpl(RootStaticContextImpl.INSTANCE);
+    StaticContext dCtx = null;
     final int ARG_DATA = 0;
     final int ARG_TYPE = 1;
 
     protected boolean processOperator(Mutable<ILogicalOperator> opRef, IOptimizationContext context)
             throws AlgebricksException {
-        AbstractLogicalOperator op1 = (AbstractLogicalOperator) opRef.getValue();;
+        VXQueryOptimizationContext vxqOptCtx = (VXQueryOptimizationContext) context;
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushChildIntoDataScanRule.java]: Seems like you could create fewer variables here. Only the dCtx is actually used later in the plan.

File: src/site/site.xml [hunk 4]
```diff
@@ -101,9 +101,9 @@ limitations under the License.
             <item
                 name="XML Data Model Example"
                 href="development_xml_document.html" />
-            <!-- <item
-                name="XQuery Benchmarks"
-                href="development_benchmarks.html" /> -->
+            <item
+                name="XQuery Benchmark Documentation"
+                href="development_benchmarks.html" />
```
Comment (prestoncarman) [inline src/site/site.xml:106]: Use the name development_xmark_status for your file name.

File: src/site/apt/development_xmark_documentation.apt [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+~~ Unless required by applicable law or agreed to in writing, software
+~~ distributed under the License is distributed on an "AS IS" BASIS,
+~~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+~~ See the License for the specific language governing permissions and
+~~ limitations under the License.
+
+
+XMark Benchmark Documentation
+    
+    This page displays VXQuery's performance on XMark Benchmark. In the following
```
Comment (prestoncarman) [inline src/site/apt/development_xmark_documentation.apt]: Remove "This page displays VXQuery's performance on XMark Benchmark."

File: src/site/apt/development_xmark_documentation.apt [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+~~ distributed under the License is distributed on an "AS IS" BASIS,
+~~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+~~ See the License for the specific language governing permissions and
+~~ limitations under the License.
+
+
+XMark Benchmark Documentation
+    
+    This page displays VXQuery's performance on XMark Benchmark. In the following 
+    table, <Query Number> column represents the XMark query number. <Original Query> column
```
Comment (prestoncarman) [inline src/site/apt/development_xmark_documentation.apt]: Add "in the following table, XMark queris are listed to show the current status of each query in VXQuery. the Query Number ...."

File: src/site/apt/development_xmark_documentation.apt [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+~~ limitations under the License.
+
+
+XMark Benchmark Documentation
+    
+    This page displays VXQuery's performance on XMark Benchmark. In the following 
+    table, <Query Number> column represents the XMark query number. <Original Query> column 
+    represents the original XMark query from the XMark Benchmark site. Original query 
+    works if it has a <Working> status in the column. If the original query does not work, then 
+    the column displays the error. <Optimized Query> column represents the optimized version
```
Comment (prestoncarman) [inline src/site/apt/development_xmark_documentation.apt]: ...displays the error "type".

Comment (prestoncarman) [inline src/site/apt/development_xmark_documentation.apt]: <Optimized Query> column represents the parallel optimized for the original XMark query.

File: src/site/apt/development_xmark_documentation.apt [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    
+    This page displays VXQuery's performance on XMark Benchmark. In the following 
+    table, <Query Number> column represents the XMark query number. <Original Query> column 
+    represents the original XMark query from the XMark Benchmark site. Original query 
+    works if it has a <Working> status in the column. If the original query does not work, then 
+    the column displays the error. <Optimized Query> column represents the optimized version 
+    of the original XMark query. The optimized version of the query 
+    has been written to increase performance. Optimized query 
+    works if it has a <Working> status in the column. If the optimized query does not work, then 
+    the column displays the error or displays <Same Error as Original Query>.
```
Comment (prestoncarman) [inline src/site/apt/development_xmark_documentation.apt]: error type ...

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushChildIntoDataScanRule.java [hunk 9]
```diff
@@ -120,10 +118,15 @@ private boolean updateDataSource(VXQueryCollectionDataSource ds, Mutable<ILogica
         ExpressionToolbox.findAllFunctionExpressions(expression, BuiltinOperators.CHILD.getFunctionIdentifier(), finds);
         for (int i = finds.size(); i > 0; --i) {
             int typeId = ExpressionToolbox.getTypeExpressionTypeArgument(finds.get(i - 1));
+            SequenceType sType = null;
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushChildIntoDataScanRule.java]: I don't think you need to declare the variable here. Either move it to be a class variable or move it inside the if statement.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushChildIntoDataScanRule.java [hunk 10]
```diff
@@ -72,13 +66,17 @@
...
     final int ARG_TYPE = 1;
 
     protected boolean processOperator(Mutable<ILogicalOperator> opRef, IOptimizationContext context)
             throws AlgebricksException {
-        AbstractLogicalOperator op1 = (AbstractLogicalOperator) opRef.getValue();;
+        VXQueryOptimizationContext vxqOptCtx = (VXQueryOptimizationContext) context;
+        IMetadataProvider iProvider = (IMetadataProvider) vxqOptCtx.getMetadataProvider();
+        VXQueryMetadataProvider vxqProvider = (VXQueryMetadataProvider) iProvider;
+        dCtx = vxqProvider.getStaticContext();
+        AbstractLogicalOperator op1 = (AbstractLogicalOperator) opRef.getValue();
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushChildIntoDataScanRule.java:77]: I think you could also conditionally set the dCtx when its null. After that it should be static.

**Comment by prestoncarman** [comment]
The review is now done. Let me know once you made your changes.

**Comment by shivani1494** [comment]
I made the changes.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushChildIntoDataScanRule.java (hunk 9 above)
Comment (shivanimall) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushChildIntoDataScanRule.java]: The latest commit does not have this change. The next commit that I make will have this one.

**Comment by shivanimall** [comment]
code updated. Thanks!

**Comment by shivani1494** [comment]
Oops, I changed the location of static context not sequence type. making a new commit for this.

**Comment by prestoncarman** [comment]
I made a few changes to your branch and then committed it to master.
- index.html had not been updated to point to the new file.
- streamlined some code in the rewrite rule.
- synced the result files between XMark original and modified.

**Pull request closed without merging**